components/encoder/ams/ams_as5048.go

```diff
@@ -125,59 +122,80 @@ func (cfg *I2CAttrConfig) ValidateI2C(path string) error {
 // Encoder is a struct representing an instance of a hardware unit
 // in AMS's AS5048 series of Hall-effect encoders.
 type Encoder struct {
+	resource.Named
 	mu                      sync.RWMutex
 	logger                  golog.Logger
 	position                float64
 	positionOffset          float64
 	rotations               int
-	connectionType          string
 	positionType            encoder.PositionType
 	i2cBus                  board.I2C
 	i2cAddr                 byte
 	cancelCtx               context.Context
 	cancel                  context.CancelFunc
 	activeBackgroundWorkers sync.WaitGroup
-	generic.Unimplemented
 }
 
 func newAS5048Encoder(
-	ctx context.Context, deps registry.Dependencies,
-	cfg config.Component, logger *zap.SugaredLogger,
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+	logger *zap.SugaredLogger,
 ) (encoder.Encoder, error) {
-	attr, ok := cfg.ConvertedAttributes.(*AttrConfig)
-	if !ok {
-		return nil, rdkutils.NewUnexpectedTypeError(attr, cfg.ConvertedAttributes)
-	}
 	cancelCtx, cancel := context.WithCancel(ctx)
 	res := &Encoder{
-		connectionType: attr.ConnectionType,
-		cancelCtx:      cancelCtx,
-		cancel:         cancel,
-		logger:         logger,
-		positionType:   encoder.PositionTypeTICKS,
+		Named:        conf.ResourceName().AsNamed(),
+		cancelCtx:    cancelCtx,
+		cancel:       cancel,
+		logger:       logger,
+		positionType: encoder.PositionTypeTicks,
 	}
-	brd, err := board.FromDependencies(deps, attr.BoardName)
-	if err != nil {
+	if err := res.Reconfigure(ctx, deps, conf); err != nil {
 		return nil, err
 	}
+	if err := res.startPositionLoop(ctx); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Reconfigure reconfigures the encoder atomically.
+func (enc *Encoder) Reconfigure(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+) error {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+
+	brd, err := board.FromDependencies(deps, newConf.BoardName)
+	if err != nil {
+		return err
+	}
 	localBoard, ok := brd.(board.LocalBoard)
 	if !ok {
-		return nil, errors.Errorf(
-			"board with name %s does not implement the LocalBoard interface", attr.BoardName,
+		return errors.Errorf(
+			"board with name %s does not implement the LocalBoard interface", newConf.BoardName,
 		)
 	}
-	if res.connectionType == i2cConn {
-		i2c, exists := localBoard.I2CByName(attr.I2CBus)
-		if !exists {
-			return nil, errors.Errorf("unable to find I2C bus: %s", attr.I2CBus)
-		}
-		res.i2cBus = i2c
-		res.i2cAddr = byte(attr.I2CAddr)
+	if newConf.ConnectionType != i2cConn {
```

Comment:
We could but out of scope here. This is just cleanup

Reply:
the thing is, not being i2c is handled by validate, and I'd rather not change this driver because it was clearer what needed to be added when spi comes along without a secondary check lower down.